pymeasure/instruments/oxfordinstruments/ips120_10.py

```diff
@@ -0,0 +1,330 @@
+    )
+
+    current_setpoint = Instrument.control(
+        "R0", "$I%f",
+        """ A floating point property that controls the magnet current set-point of
+        the IPS in ampere. """,
+        get_process=lambda v: float(v[1:]),
+        validator=truncated_range,
+        values=[0, _MAX_CURRENT]
+    )
+
+    field_setpoint = Instrument.control(
+        "R8", "$J%f",
+        """ A floating point property that controls the magnetic field set-point of
+        the IPS in Tesla. """,
+        get_process=lambda v: float(v[1:]),
+        # validator=truncated_range,
+        # values=_T_RANGE
+    )
+
+    sweep_rate = Instrument.control(
+        "R9", "$T%f",
+        """ A floating point property that controls the sweep-rate of
+        the IPS in Tesla/minute. """,
+        get_process=lambda v: float(v[1:]),
+        # validator=truncated_range,
+        # values=_T_RANGE
+    )
+
+    activity = Instrument.control(
+        "X", "$A%d",
+        """ A string property that controls the activity of the IPS. Valid values
+        are "hold", "to setpoint", "to zero" and "clamp" """,
+        get_process=lambda v: int(v[4]),
+        values={"hold": 0, "to setpoint": 1, "to zero": 2, "clamp": 4},
+        map_values=True,
+    )
+
+    sweep_status = Instrument.measurement(
+        "X",
+        """ A string property that returns the current sweeping mode of
+        the IPS. """,
+        get_process=lambda v: int(v[11]),
+        values={"at rest": 0, "sweeping": 1, "sweep limiting": 2, "sweeping & sweep limiting": 3},
+        map_values=True,
+    )
+
+    @property
+    def field(self):
+        """ Property that returns the current magnetic field value in Tesla.
+        """
+
+        sw_heater = self.switch_heater_status
+
+        if sw_heater in [0, 2]:
+            field = self.persistent_field
+        elif sw_heater in [1]:
+            field = self.demand_field
+        else:
+            log.error("IPS: Switch status returned %d" % sw_heater)
+            field = self.demand_field
+
+        return field
+
+    def enable_control(self):
+        """ Method that enables active control of the IPS.
+        Sets control to remote and turns off the clamp. """
+        self.control_mode = "RU"
+
+        # Turn off clamping if still clamping
+        if self.activity == "clamp":
+            self.activity = "hold"
+
+        # Turn on switch-heater if field at zero
+        if self.field == 0:
+            self.switch_heater_status = 1
+
+    def disable_control(self):
+        if not self.field == 0:
+            raise Exception("IPS not at 0T, not disabling the supply")
+
+        self.switch_heater_status = 0
+        self.activity = "clamp"
+
+    def enable_persistent_mode(self):
+        """ Methods that enables the persistent magnetic field mode. """
+        # Check if system idle
+        if not self.sweep_status == "at rest":
+            raise Exception("Magnet not at rest")
```

Comment:
Might make sense to create custom exceptions in this module (`MagnetError`, and `SwitchStatusError`, deriving from `ValueError`?) for easier handling. Otherwise, whoever handles these exceptions would need to check for the exception message to tell this apart from, e.g. a `TimeoutError`.

Reply:
similar for the next method

Reply:
Makes perfect sense indeed; when I implemented it I simply did not have time for a nice implementation (just needed it to work properly since we had only access to the magnet for about a week).